opm/grid/LookUpData.hh

```diff
@@ -261,131 +255,186 @@ protected:
-    return feature_vec[this->getOriginIndexFromEntity<EntityType,Grid>(elem)];
+    const auto& fieldPropIdx = this->getFieldPropIdx<Grid>(elemIdx);
+    assert(0 <= fieldPropIdx && static_cast<int>(fieldProp.size()) > fieldPropIdx);
+    return fieldProp[fieldPropIdx];
 }
 
 template<typename Grid, typename GridView>
-template<typename FeatureType>
-FeatureType Opm::LookUpData<Grid,GridView>::operator()(const int& elemIdx, const std::vector<FeatureType>& feature_vec) const
+template<typename ElemOrIndex>
+double Opm::LookUpData<Grid,GridView>::fieldPropDouble(const FieldPropsManager& fieldPropsManager,
+                                                       const std::string& propString,
+                                                       const ElemOrIndex& elemOrIndex) const
 {
-    assert(0 <= this-> getOriginIndex<Grid>(elemIdx) && static_cast<int>(feature_vec.size()) > this-> getOriginIndex<Grid>(elemIdx));
-    return feature_vec[getOriginIndex<Grid>(elemIdx)];
+    const auto& fieldPropVec = fieldPropsManager.get_double(propString);
+    return this ->operator()(elemOrIndex,fieldPropVec);
+}
+
+template<typename Grid, typename GridView>
+template<typename ElemOrIndex>
+int Opm::LookUpData<Grid,GridView>::fieldPropInt(const FieldPropsManager& fieldPropsManager,
+                                                 const std::string& propString,
+                                                 const ElemOrIndex& elemOrIndex) const
+{
+    const auto& fieldPropVec = fieldPropsManager.get_int(propString);
+    return this ->operator()(elemOrIndex,fieldPropVec);
 }
 
 template<typename Grid, typename GridView>
 template<typename EntityType, typename GridType>
 typename std::enable_if_t<!std::is_same_v<GridType,Dune::CpGrid>,int>
-Opm::LookUpData<Grid,GridView>::getOriginIndexFromEntity(const EntityType& elem) const
+Opm::LookUpData<Grid,GridView>::getFieldPropIdx(const EntityType& elem) const
 {
     static_assert(std::is_same_v<Grid,GridType>);
+    assert(elem.level() == 0); // LGRs (level>0) only supported for CpGrid.
     return this-> elemMapper_.index(elem);
 }
 
 template<typename Grid, typename GridView>
-template<typename EntityType, typename GridType>
+template<typename EntityType,typename GridType>
 typename std::enable_if_t<std::is_same_v<GridType,Dune::CpGrid>,int>
-Opm::LookUpData<Grid,GridView>::getOriginIndexFromEntity(const EntityType& elem) const
+Opm::LookUpData<Grid,GridView>::getFieldPropIdx(const EntityType& elem) const
 {
     static_assert(std::is_same_v<Grid,GridType>);
-    return elem.getOrigin().index();
+    static_assert(std::is_same_v<EntityType,Dune::cpgrid::Entity<0>>);
+    if (isFieldPropInLgr_ && elem.level()) { // level > 0 == true ; level == 0 == false
+        // In case some LGRs do not have refined field properties, the next line need to be modified.
+        return elem.getLevelElem().index(); // getLgrElem() returns equivalent refined Entity in the LGR the leafGridView cell was born.
```

Comment:
Not a big one, but the method name used in comment is wrong now. Maybe just skip it?

Reply:
Skipped